src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,359 @@
+			// servers in s.servers won't actually serve anymore, and instead parent
+			// server will forward requests based on the request hostname
+			port := s.apiServer.env.S3GatewayPort
+			server, err := s3.Server(port, driver, s)
+			if err != nil {
+				return fmt.Errorf("sidecar s3 gateway: couldn't initialize s3 gateway server: %v", err)
+			}
+			strport := strconv.FormatInt(int64(port), 10)
+			listener, err := net.Listen("tcp", ":"+strport)
+			if err != nil {
+				return fmt.Errorf("sidecar s3 gateway: could not serve on port %q: %v", port, err)
+			}
+			go func() {
+				server.Serve(listener)
+			}()
+			s.servers[jobID] = server
+			return nil
+		},
+
+		End: func(jobID string) error {
+			s.serversMu.Lock()
+			defer s.serversMu.Unlock()
+			// kill server
+			server, ok := s.servers[jobID]
+			if !ok {
+				// Note that because we call InspectJob after receiving a creation event, we
+				// may never create a service for a job that is created and then immediately
+				// deleted, so no error is returned if jobID is not in s.servers
+				return nil
+			}
+			if err := server.Close(); err != nil {
+				return fmt.Errorf("could not kill sidecar s3 gateway server for job %q: %v", jobID, err)
+			}
+			delete(s.servers, jobID)
+			return nil
+		},
+	})
+
+	return fmt.Errorf("sidecar s3 gateway: Serve() is exiting, which shouldn't happen")
+}
+
+func (s *sidecarS3G) master() (retErr error) {
+	logrus.Infof("Launching sidecar s3 gateway master process")
+	defer func() {
+		panic(
+			fmt.Sprintf("sidecar s3 gateway master is exiting; this should never happen (err: %v)", retErr),
+		)
+	}()
+	masterLock := dlock.NewDLock(s.apiServer.env.GetEtcdClient(),
+		path.Join(s.apiServer.etcdPrefix,
+			s3gSidecarLockPath,
+			s.pipelineInfo.Pipeline.Name,
+			s.pipelineInfo.Salt))
+	ctx, err := masterLock.Lock(s.pachClient.Ctx())
+	if err != nil {
+		return err
+	}
+	// pachClient := s.pachClient.WithCtx(ctx)
+	defer masterLock.Unlock(ctx)
+	// Watch for new jobs & create kubernetes service for each new job
+	s.HandleJobs(JobHandler{
+		Begin: func(ctx context.Context, jobInfo *pps.JobInfo) error {
+			// Create kubernetes service for the current job ('jobInfo')
+			rcName := ppsutil.PipelineRcName(jobInfo.Pipeline.Name, jobInfo.PipelineVersion)
+			labels := map[string]string{
+				"app":       rcName,
+				"suite":     "pachyderm",
+				"component": "worker",
+			}
+			service := &v1.Service{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Service",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   rcName,
+					Labels: labels,
+				},
+				Spec: v1.ServiceSpec{
+					Selector: labels,
+					Ports: []v1.ServicePort{
+						{
+							Port: int32(s.apiServer.env.S3GatewayPort),
+							Name: "s3-gateway-port",
+						},
+					},
+				},
+			}
+
+			if err := backoff.RetryNotify(func() error {
+				_, err := s.apiServer.env.GetKubeClient().CoreV1().Services(s.apiServer.namespace).Create(service)
```

Comment:
Is it fair to assume that the kube account can create/delete services? I assume so since presumably the same account coordinates pipeline RC creation/deletion, but wanted to double-check.

Reply:
It's a great question, and fortunately the answer is yes: part of this PR is to create, in `pachctl deploy`, a service account with a service-creation-allowing role that worker pods are now assigned on creation and will use to do this.

It's significantly more restricted than the `pachyderm` service account that the `pachd` pods use, and is basically limited to being able to do this.